server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -2282,17 +2285,26 @@ public void deletePools(Collection<Pool> pools, Collection<String> alreadyDelete
                     this.enforcer.postUnbind(entitlement.getConsumer(), this, entitlement);
                 }
 
-                // Recalculate status for affected consumers
-                log.info("Recomputing status for {} consumers", consumerStackedEnts.size());
-                int i = 0;
-                for (Consumer consumer : consumerStackedEnts.keySet()) {
-                    this.complianceRules.getStatus(consumer);
+                /* Save the consumer IDs and then detach all the keys in the consumerStackEnts map.
+                 * Otherwise during the status calculations, the facts proxy objects objects will be resolved
+                 * and the memory use will grow linearly with the number of consumers instead of remaining
+                 * constant as we calculate the status of each consumer.  See BZ 1584259 */
+                List<String> consumers = consumerStackedEnts.keySet().stream()
+                    .map(Consumer::getId)
+                    .collect(Collectors.toList());
+                this.consumerCurator.batchDetach(consumerStackedEnts.keySet());
```

Comment:
I'm still a little fuzzy on the particulars of the streaming API, but if I understand it correctly, we can avoid a second iteration through the collection by adding a forEach before the map to do the detaching:
```
List<String> consumers = ...
    .forEach(c -> this.consumerCurator.detach(c))
    .map(Consumer::getId)
    ...
```

My intent here is to do the detach and ID fetching in-line, rather than as two separate iterations through the collection.

As a side note, I just learned that we can apparently reference functions via the double colon syntax instead of manually calling them? Does this mean in my above example, we could just as easily do ```ConsumerCurator::detach``` rather than the longer lambda syntax? Feels like the scoping is wrong for such a ref.

Reply:
That doesn't work.  `forEach` is a [void function](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#forEach-java.util.function.Consumer-).  It could be done by overloading the map function such that it detached the item and then grabbed the id, but that's adding a side-effect into what is meant to be functional code for a time savings that's probably measured in milliseconds.  Not worth it in my opinion.